src/command_line_parser.cc

```diff
@@ -0,0 +1,1528 @@
+// Copyright 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+#include "command_line_parser.h"
+
+#include <iomanip>
+#include <iostream>
+#include <string>
+
+namespace triton { namespace server {
+
+// [FIXME] expose following parse helpers for other type of parser
+namespace {
+
+bool
+ParseBoolOption(std::string arg)
+{
+  std::transform(arg.begin(), arg.end(), arg.begin(), [](unsigned char c) {
+    return std::tolower(c);
+  });
+
+  if ((arg == "true") || (arg == "on") || (arg == "1")) {
+    return true;
+  }
+  if ((arg == "false") || (arg == "off") || (arg == "0")) {
+    return false;
+  }
+
+  throw ParseException("invalid value for bool option: " + arg);
+}
+
+// Template specialization for ParsePairOption
+// [FIXME] replace ParseXXXOPtion with these
+template <typename T>
+T ParseOption(const std::string& arg);
+
+template <>
+int
+ParseOption(const std::string& arg)
+{
+  return std::stoi(arg);
+}
+
+template <>
+uint64_t
+ParseOption(const std::string& arg)
+{
+  return std::stoll(arg);
+}
+
+int
+ParseIntOption(const std::string arg)
+{
+  return std::stoi(arg);
+}
+
+int64_t
+ParseLongLongOption(const std::string arg)
+{
+  return std::stoll(arg);
+}
+
+#if 0
+float
+ParseFloatOption(const std::string arg)
+{
+  return std::stof(arg);
+}
+#endif
+
+template <>
+double
+ParseOption(const std::string& arg)
+{
+  return std::stod(arg);
+}
+
+// Condition here merely to avoid compilation error, this function will
```

Comment:
Optional if you agree, but I believe this file could be a good candidate for starting to break out some ifdefs into `constexpr bool`s like [rapids_triton does](https://github.com/rapidsai/rapids-triton/blob/4a522cfacf9d1792376056c30d63657ee34b9d91/cpp/include/rapids_triton/build_control.hpp#L24-L28). Then all `ifdefs` in this source can be replaced with `if (logging_enabled) { }` and it shouldn't complain about unused functions/variables even when these are disabled.

Reply:
This is going to be a different approach to address the problem, more on runtime check (and compiler optimization) rather than compile time selection. I am not going to do this in this PR and it will need a discussion before we agree to move towards this way.

Reply:
Sounds good.